File: 3967-earliest-finish-time-for-land-and-water-rides-ii/earliest-finish-time-for-land-and-water-rides-ii.py
from bisect import bisect_right
from typing import List

class Solution:
    def solve(self, firstStart, firstDuration,
              secondStart, secondDuration):

        rides = sorted(zip(secondStart, secondDuration))

        starts = [s for s, d in rides]

        m = len(rides)

        prefix_min_duration = [0] * m
        prefix_min_duration[0] = rides[0][1]

        for i in range(1, m):
            prefix_min_duration[i] = min(
                prefix_min_duration[i - 1],
                rides[i][1]
            )

        suffix_min_finish = [0] * m
        suffix_min_finish[-1] = rides[-1][0] + rides[-1][1]

        for i in range(m - 2, -1, -1):
            suffix_min_finish[i] = min(
                suffix_min_finish[i + 1],
                rides[i][0] + rides[i][1]
            )

        ans = float('inf')

        for start, duration in zip(firstStart, firstDuration):

            finish_first = start + duration

            pos = bisect_right(starts, finish_first) - 1

            if pos >= 0:
                ans = min(
                    ans,
                    finish_first + prefix_min_duration[pos]
                )

            if pos + 1 < m:
                ans = min(
                    ans,
                    suffix_min_finish[pos + 1]
                )

        return ans

    def earliestFinishTime(
        self,
        landStartTime: List[int],
        landDuration: List[int],
        waterStartTime: List[int],
        waterDuration: List[int]
    ) -> int:

        ans1 = self.solve(
            landStartTime,
            landDuration,
            waterStartTime,
            waterDuration
        )

        ans2 = self.solve(
            waterStartTime,
            waterDuration,
            landStartTime,
            landDuration
        )

        return min(ans1, ans2)

File: 3967-earliest-finish-time-for-land-and-water-rides-ii/earliest-finish-time-for-land-and-water-rides-ii.java
import java.util.*;

class Solution {

    private long solve(int[] firstStart, int[] firstDuration,
                       int[] secondStart, int[] secondDuration) {

        int m = secondStart.length;

        int[][] rides = new int[m][2];
        for (int i = 0; i < m; i++) {
            rides[i][0] = secondStart[i];
            rides[i][1] = secondDuration[i];
        }

        Arrays.sort(rides, Comparator.comparingInt(a -> a[0]));

        int[] starts = new int[m];
        long[] prefixMinDuration = new long[m];
        long[] suffixMinFinish = new long[m];

        for (int i = 0; i < m; i++) {
            starts[i] = rides[i][0];
        }

        prefixMinDuration[0] = rides[0][1];
        for (int i = 1; i < m; i++) {
            prefixMinDuration[i] =
                Math.min(prefixMinDuration[i - 1], rides[i][1]);
        }

        suffixMinFinish[m - 1] =
            (long) rides[m - 1][0] + rides[m - 1][1];

        for (int i = m - 2; i >= 0; i--) {
            long val = (long) rides[i][0] + rides[i][1];
            suffixMinFinish[i] =
                Math.min(suffixMinFinish[i + 1], val);
        }

        long ans = Long.MAX_VALUE;

        for (int i = 0; i < firstStart.length; i++) {

            long finishFirst =
                (long) firstStart[i] + firstDuration[i];

            int pos = upperBound(starts, (int) finishFirst);

            if (pos >= 0) {
                ans = Math.min(
                    ans,
                    finishFirst + prefixMinDuration[pos]
                );
            }

            if (pos + 1 < m) {
                ans = Math.min(
                    ans,
                    suffixMinFinish[pos + 1]
                );
            }
        }

        return ans;
    }

    private int upperBound(int[] arr, int target) {
        int l = 0, r = arr.length;

        while (l < r) {
            int mid = l + (r - l) / 2;

            if (arr[mid] <= target) {
                l = mid + 1;
            } else {
                r = mid;
            }
        }

        return l - 1;
    }

    public int earliestFinishTime(int[] landStartTime, int[] landDuration,
                                  int[] waterStartTime, int[] waterDuration) {

        long ans1 = solve(
            landStartTime, landDuration,
            waterStartTime, waterDuration
        );

        long ans2 = solve(
            waterStartTime, waterDuration,
            landStartTime, landDuration
        );

        return (int) Math.min(ans1, ans2);
    }
}